### PR TITLE
Update integration tests to use REST APIs in master branch (II)

### PR DESCRIPTION
**Pull request by ashanthamara**
### OAuth2
- Oauth2TokenRenewalPerRequestTestCase 
- SystemScopePermissionValidationTestCase
- Oauth2PersistenceProcessorTestCase
- Oauth2PersistenceProcessorInsertTokenTestCase
- Oauth2HashAlgorithmTestCase
- OAuth2TokenScopeValidatorTestCase
- OAuth2TokenRevocationWithSessionTerminationTestCase
- OAuth2ServiceAuthCodeGrantTestCase
- OAuth2ServiceAuthCodeGrantOpenIdTestCase
- OAuth2ServiceAuthCodeGrantOpenIdRequestObjectTestCase
- OAuth2ServiceJWTGrantTestCase
- OIDCCustomScopesLoginTest

## Added following rest clients
- `ClaimManagementRestClient` 
-  `ChallengeQuestionsRestClient`
-  `OIDCScopeMgtRestClient`
-  `IdpMgtRestClient`

## Related Issue

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/model/IdentityProviderPOSTRequest.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        @Override
+        public String toString() {
+
+            return "class Certificate {\n" +
+                    "    certificates: " + toIndentedString(certificates) + "\n" +
+                    "    jwksUri: " + toIndentedString(jwksUri) + "\n" +
+                    "}";
+        }
+    }
+}
```
Comment (madurangasiriwardena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/model/IdentityProviderPOSTRequest.java:400]: add new line

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java [hunk 2]
```diff
@@ -48,21 +53,24 @@ public void testInit() throws Exception {
...
+        Assert.assertNotNull(application, "Application creation failed.");
+        String applicationId1 = application.getId();
 
-        consumerKey1 = oAuthConsumerAppDTO.getOauthConsumerKey();
+        OpenIDConnectConfiguration oidcConfig = getOIDCInboundDetailsOfApplication(applicationId1);
+        String consumerKey1 = oidcConfig.getClientId();
         Assert.assertNotNull(consumerKey1, "Application creation failed.");
 
-        String consumerSecret1 = oAuthConsumerAppDTO.getOauthConsumerSecret();
+        String consumerSecret1  = oidcConfig.getClientSecret();
```
Comment (madurangasiriwardena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java:70]:
```suggestion
        String consumerSecret1 = oidcConfig.getClientSecret();
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java [hunk 3]
```diff
@@ -72,41 +80,45 @@ public void testEncryptionDecryptionPersistenceProcessor() throws Exception {
...
+        Assert.assertNotNull(application, "Application creation failed.");
+        String applicationId2 = application.getId();
 
-        consumerKey2 = oAuthConsumerAppDTO.getOauthConsumerKey();
+        OpenIDConnectConfiguration oidcConfig = getOIDCInboundDetailsOfApplication(applicationId2);
+        String consumerKey2 = oidcConfig.getClientId();
         Assert.assertNotNull(consumerKey2, "Application creation failed.");
 
-        String consumerSecret2 = oAuthConsumerAppDTO.getOauthConsumerSecret();
+        String consumerSecret2  = oidcConfig.getClientSecret();
```
Comment (madurangasiriwardena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java:91]:
```suggestion
        String consumerSecret2 = oidcConfig.getClientSecret();
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2HashAlgorithmTestCase.java [hunk 4]
```diff
@@ -247,9 +246,9 @@ public void testValidateAccessToken() throws Exception {
 
     private void registerAndDeployApplication() throws Exception {
```
Comment (madurangasiriwardena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2HashAlgorithmTestCase.java:247]: This method is not deploying any application, right? Shall we rename it?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/model/IdentityProviderPOSTRequest.java (hunk 1 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/idp/v1/model/IdentityProviderPOSTRequest.java]: Fixed with https://github.com/wso2/product-is/pull/16271/commits/e9343598362e5021e1f6832b61ef1997495afeda

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java (hunk 2 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java]: Fixed with https://github.com/wso2/product-is/pull/16271/commits/e9343598362e5021e1f6832b61ef1997495afeda

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java (hunk 3 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2PersistenceProcessorTestCase.java]: Fixed with https://github.com/wso2/product-is/pull/16271/commits/e9343598362e5021e1f6832b61ef1997495afeda

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2HashAlgorithmTestCase.java (hunk 4 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/Oauth2HashAlgorithmTestCase.java]: Fixed with https://github.com/wso2/product-is/pull/16271/commits/e9343598362e5021e1f6832b61ef1997495afeda
Changed the method name to `createApp`

**Review by madurangasiriwardena — APPROVED** [review]

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java [hunk 5]
```diff
@@ -93,9 +96,8 @@ public class OAuth2ServiceAbstractIntegrationTest extends ISIntegrationTest {
 	/**
 	 * Initialize
 	 *
-	 * @param userMode
-	 *            - User Id
-	 * @throws Exception
+	 * @param userMode - User Id
+	 * @throws Exception Exception
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:100]: Shall we fix the formatting by adding missing full stops? Le's check other places also

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java [hunk 6]
```diff
@@ -308,49 +334,57 @@ public OAuthConsumerAppDTO createApplication(OAuthConsumerAppDTO appDTO, String
...
-	 * @param url
-	 * @return
-	 * @throws ClientProtocolException
-	 * @throws java.io.IOException
+	 * @param client HttpClient
+	 * @param urlParameters url parameters
+	 * @param url endpoint
+	 * @return HttpResponse
+	 * @throws ClientProtocolException ClientProtocolException
+	 * @throws java.io.IOException java.io.IOException
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:384]: Shall we fix the formatting by adding missing fullstops?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdRequestObjectTestCase.java [hunk 7]
```diff
@@ -344,80 +363,46 @@ public void testClaims() throws Exception {
...
-        Assert.assertNull(((org.json.simple.JSONObject) obj).get("externalClaim1"));
-        Assert.assertNull(((org.json.simple.JSONObject) obj).get("externalClaim2"));
+        Assert.assertNull(((JSONObject) obj).get("externalClaim1"));
+        Assert.assertNull(((JSONObject) obj).get("externalClaim2"));
     }
 
     @Test(groups = "wso2.is", description = "Validate Token Expiration Time",
             dependsOnMethods = "testValidateAccessToken")
     public void testValidateTokenExpirationTime() throws Exception {
-        OAuth2TokenValidationRequestDTO requestDTO = new OAuth2TokenValidationRequestDTO();
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdRequestObjectTestCase.java:361]: Shall we add the missing new line. Let's check in other places as well.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdTestCase.java [hunk 8]
```diff
@@ -317,59 +306,25 @@ public void testClaims() throws Exception {
...
-            Assert.assertTrue(jwtClaimMappingRoleValues.contains(","), "Broken JWT Token from Authorization context");
-
-            String[] jwtClaimMappingRoleElements = jwtClaimMappingRoleValues.replaceAll("[\\[\\]\"]", "").split(",");
-            List<String> jwtClaimMappingRoleElementsList = Arrays.asList(jwtClaimMappingRoleElements);
-            Assert.assertTrue(jwtClaimMappingRoleElementsList.contains("Internal/everyone"), "Invalid JWT Token Role " +
-                    "Values");
-        }
+            "testValidateAccessToken")
+    public void testValidateTokenScope() throws Exception {
+        JSONObject tokenResponse = introspectToken();
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdTestCase.java:320]: Shall we add the missing new line?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceJWTGrantTestCase.java [hunk 9]
```diff
@@ -43,108 +45,128 @@
...
+        oidcScopeMgtRestClient =  new OIDCScopeMgtRestClient(serverURL, tenantInfo);
+        idpMgtRestClient = new IdpMgtRestClient(serverURL, tenantInfo);
+
+        addAdminUser();
+        changeCountryOIDCDialect();
     }
 
     @AfterClass(alwaysRun = true)
     public void atEnd() throws Exception {
+        deleteApp(applicationId);
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceJWTGrantTestCase.java:153]: Shall we add the missing new line here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2TokenScopeValidatorTestCase.java [hunk 10]
```diff
@@ -222,4 +222,38 @@ private TokenIntrospectionResponse introspectAccessToken(AccessToken accessToken
 
         return TokenIntrospectionResponse.parse(introspectionHTTPResp);
     }
+
+    private void createOAuthApplication() throws Exception {
+        restClient = new OAuth2RestClient(serverURL, tenantInfo);
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2TokenScopeValidatorTestCase.java:227]: Shall we add the missing new line here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/SCIM2RestClient.java [hunk 11]
```diff
@@ -104,7 +104,23 @@ public void updateUser(PatchOperationRequestObject patchUserInfo, String userId)
 
         try (CloseableHttpResponse response = getResponseOfHttpPatch(endPointUrl, jsonRequest, getHeaders())) {
             Assert.assertEquals(response.getStatusLine().getStatusCode(), HttpServletResponse.SC_OK,
-                    "Role update failed");
+                    "User update failed");
+        }
+    }
+
+    /**
+     * Search a user and get requested attributes
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/SCIM2RestClient.java:112]: Shall we add the missing full stop here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/OIDCScopeMgtRestClient.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+import org.testng.Assert;
+import org.wso2.carbon.automation.engine.context.beans.Tenant;
+import org.wso2.identity.integration.test.rest.api.server.oidc.scope.management.v1.model.ScopeUpdateRequest;
+import org.wso2.identity.integration.test.utils.OAuth2Constant;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class OIDCScopeMgtRestClient extends RestBaseClient {
+    private final String serverUrl;
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/OIDCScopeMgtRestClient.java:39]: Shall we add a new line here?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/IdpMgtRestClient.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        this.password = tenantInfo.getContextUser().getPassword();
+    }
+
+    /**
+     * Create an Identity Provider.
+     *
+     * @param idpCreateReqObj Identity Provider request object.
+     */
+    public String createIdentityProvider(IdentityProviderPOSTRequest idpCreateReqObj) throws Exception {
+        String jsonRequest = toJSONString(idpCreateReqObj);
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/IdpMgtRestClient.java:59]: Shall we add the missing new line here?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ClaimManagementRestClient.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    private static final String CLAIM_DIALECTS_ENDPOINT_URI = "/claim-dialects";
+
+    public static final String CLAIMS_ENDPOINT_URI = "/claims";
+    private final CloseableHttpClient client;
+    private final String username;
+    private final String password;
+    private final String serverBasePath;
+
+    public ClaimManagementRestClient(String backendURL, Tenant tenantInfo) {
+        client = HttpClients.createDefault();
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ClaimManagementRestClient.java:50]: Shall we add the missing new line here?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ChallengeQuestionsRestClient.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+import org.testng.Assert;
+import org.wso2.carbon.automation.engine.context.beans.Tenant;
+import org.wso2.identity.integration.test.rest.api.server.challenge.v1.model.UserChallengeAnswer;
+import org.wso2.identity.integration.test.utils.OAuth2Constant;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class ChallengeQuestionsRestClient extends RestBaseClient {
+    private final String serverUrl;
```
Comment (UdeshAthukorala) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ChallengeQuestionsRestClient.java:35]: Shall we add the missing new line here?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java (hunk 5 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:100]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java (hunk 6 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:384]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdRequestObjectTestCase.java (hunk 7 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdRequestObjectTestCase.java:361]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceJWTGrantTestCase.java (hunk 9 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceJWTGrantTestCase.java:153]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdTestCase.java (hunk 8 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAuthCodeGrantOpenIdTestCase.java:320]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2TokenScopeValidatorTestCase.java (hunk 10 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2TokenScopeValidatorTestCase.java:227]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/OIDCScopeMgtRestClient.java (hunk 12 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/OIDCScopeMgtRestClient.java:39]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/SCIM2RestClient.java (hunk 11 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/SCIM2RestClient.java:112]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/IdpMgtRestClient.java (hunk 13 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/IdpMgtRestClient.java:59]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ClaimManagementRestClient.java (hunk 14 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ClaimManagementRestClient.java:50]: Fixed with https://github.com/wso2/product-is/pull/20096

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ChallengeQuestionsRestClient.java (hunk 15 above)
Comment (ashanthamara) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/restclients/ChallengeQuestionsRestClient.java:35]: Fixed with https://github.com/wso2/product-is/pull/20096

**Pull request merged**